crates/starknet-os-types/src/starknet_core_addons.rs

```diff
@@ -0,0 +1,141 @@
+            inputs: entry.inputs,
+            name: entry.name,
+            outputs: entry.outputs,
+        }),
+    }
+}
+
+fn raw_legacy_member_from_legacy_struct_member(member: LegacyStructMember) -> RawLegacyMember {
+    RawLegacyMember { name: member.name, offset: member.offset, r#type: member.r#type }
+}
+
+/// Implementation of From<LegacyContractAbiEntry> for RawLegacyAbiEntry,
+/// as it is missing from starknet-rs and we need it to interpret compressed legacy contracts.
+fn raw_legacy_abi_entry_from_legacy_contract_abi_entry(
+    legacy_contract_abi_entry: LegacyContractAbiEntry,
+) -> RawLegacyAbiEntry {
+    match legacy_contract_abi_entry {
+        LegacyContractAbiEntry::Function(entry) => raw_abi_entry_from_legacy_function_abi_entry(entry),
+        LegacyContractAbiEntry::Struct(entry) => RawLegacyAbiEntry::Struct(RawLegacyStruct {
+            members: entry.members.into_iter().map(raw_legacy_member_from_legacy_struct_member).collect(),
+            name: entry.name,
+            size: entry.size,
+        }),
+        LegacyContractAbiEntry::Event(entry) => {
+            RawLegacyAbiEntry::Event(RawLegacyEvent { data: entry.data, keys: entry.keys, name: entry.name })
+        }
+    }
+}
+
+fn raw_legacy_entrypoint_from_legacy_entrypoint(legacy_entry_point: LegacyContractEntryPoint) -> RawLegacyEntryPoint {
+    RawLegacyEntryPoint {
+        offset: LegacyEntrypointOffset::U64AsHex(legacy_entry_point.offset),
+        selector: legacy_entry_point.selector,
+    }
+}
+
+/// Implementation of From<LegacyEntryPointsByType> for RawLegacyEntryPoints,
+/// as it is missing from starknet-rs and we need it to interpret compressed legacy contracts.
+fn raw_legacy_entrypoints_from_legacy_entrypoints(
+    legacy_entry_points_by_type: LegacyEntryPointsByType,
+) -> RawLegacyEntryPoints {
+    RawLegacyEntryPoints {
+        constructor: legacy_entry_points_by_type
+            .constructor
+            .into_iter()
+            .map(raw_legacy_entrypoint_from_legacy_entrypoint)
+            .collect(),
+        external: legacy_entry_points_by_type
+            .external
+            .into_iter()
+            .map(raw_legacy_entrypoint_from_legacy_entrypoint)
+            .collect(),
+        l1_handler: legacy_entry_points_by_type
+            .l1_handler
+            .into_iter()
+            .map(raw_legacy_entrypoint_from_legacy_entrypoint)
+            .collect(),
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum LegacyContractDecompressionError {
+    #[error("Failed to deserialize the program as JSON: {0}")]
+    Serde(#[from] serde_json::Error),
+
+    #[error("Failed to decompress the program: {0}")]
+    Decompression(#[from] std::io::Error),
+}
+
+pub(crate) fn decompress_starknet_core_contract_class(
+    compressed_legacy_class: CompressedLegacyContractClass,
+) -> Result<LegacyContractClass, LegacyContractDecompressionError> {
+    // Compressed classes have the program as a gzipped vector.
```

Comment:
This description of the problem is good but is hard to find. It seems that the `TryFrom` impl is the main entry for using this decompression, which isn't very self-descriptive.

Reply:
I improved the doc a little.